cmd/charm/charmcmd/dockerfake_test.go

```diff
@@ -99,3 +100,133 @@ func (srv *dockerHandler) addRequest(req interface{}) {
 	defer srv.mu.Unlock()
 	srv.reqs = append(srv.reqs, req)
 }
+
+func newDockerRegistryHandler(authHandlerURL string) *dockerRegistryHandler {
+	return &dockerRegistryHandler{
+		apiVersion:     "registry/2.0",
+		authHandlerURL: authHandlerURL,
+	}
+}
+
+type registeredImage struct {
+	version1 bool
+	name     string
+	digest   string
+}
+
+type dockerRegistryHandler struct {
+	apiVersion     string
+	authHandlerURL string
+	images         []*registeredImage
+	errors         []string
+}
+
+func (srv *dockerRegistryHandler) addImage(img *registeredImage) {
+	srv.images = append(srv.images, img)
+}
+
+func (srv *dockerRegistryHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	logger.Infof("dockerRegistryHandler.ServeHTTP %v", req.URL)
+	if req.Method != "GET" && req.Method != "HEAD" {
+		srv.addErrorf("unexpected method")
+		http.Error(w, "unexpected method", http.StatusMethodNotAllowed)
+		return
+	}
+	req.ParseForm()
+	if req.URL.Path == "/v2/" {
+		srv.serveV2Root(w, req)
+		return
+	}
+	parts := strings.Split(req.URL.Path, "/")
+	if len(parts) < 4 || parts[1] != "v2" || parts[len(parts)-2] != "manifests" {
+		srv.addErrorf("unexpected access to path %v", req.URL)
+		http.NotFound(w, req)
+		return
+	}
+	if req.Header.Get("Accept") != "application/vnd.docker.distribution.manifest.v2+json" {
```

Comment:
As per [1] this should include "application/vnd.docker.distribution.manifest.v2+json" & "application/vnd.docker.distribution.manifest.list.v2+json". This probably indicates that the code won't work for some images.

[1]: https://docs.docker.com/registry/spec/manifest-v2-2/#backward-compatibility

Reply:
done